src/tools/gxpkg.ss

```diff
@@ -303,16 +346,125 @@
+        (run-process ["git" "fetch" "-q"]
+                     directory: dest
+                     coprocess: void
+                     stdout-redirection: #f))
+      (let ((path (path-directory dest))
+            (clone-url (string-append "https://" pkg ".git")))
+        (displayln "... cloning " pkg)
+        (create-directory* path)
+        (run-process ["git" "clone" "-q" clone-url]
+                     directory: path
+                     coprocess: void
+                     stdout-redirection: #f)))
+    (when tag
+      (displayln "... checking out " tag)
+      (run-process ["git" "checkout" "-q" tag]
+                   directory: dest
+                   coprocess: void
+                   stdout-redirection: #f)
+      (call-with-output-file (pkg-tag-file pkg)
+        (cut write tag <>)))
+    (let (tag (pkg-tag-get pkg))
+      (when (not (pkg-tag-semver? tag))
+        (displayln "... pulling")
+        (run-process ["git" "pull" "-q" "origin" tag]
+                     directory: dest
+                     coprocess: void
+                     stdout-redirection: #f)))))
+
+(def (pkg-tag-file pkg)
+  (let* ((root (pkg-root-dir))
+         (dest (path-expand pkg root)))
+    (string-append dest ".tag")))
+
+(def (pkg-tag-get pkg)
+  (let ((tagf (pkg-tag-file pkg))
+        (top (path-expand pkg (pkg-root-dir))))
+    (cond
+     ((file-exists? tagf)
+      (call-with-input-file tagf read))
+     ((file-exists? top)
+      (run-process ["git" "branch" "--show-current"]
+                   directory: top
+                   coprocess: read-line))
+     (else #f))))
+
+(def (pkg-tag-incompatible? current other)
+  (cond
+   ((or (not current) (not other)) #f)
+   ((and (pkg-tag-semver? current)
+         (pkg-tag-semver? other))
+    #f)
+   ((or (member current '("master" "main"))
+        (member other '("master" "main")))
+    #f)
+   (else
+    (not (equal? current other)))))
+
+;; Note: in this implementation of semver, we always keep the greatest version.
+;; We don't pay attention to majors and we consider master/main to be the frontier.
+(def (pkg-tag-choose current other)
+  (cond
+   ((equal? current other)
+    current)
+   ((not other)
+    current)
+   ((member current '("master" "main"))
+    current)
+   ((member other '("master" "main"))
+    other)
+   (else
+    (let ((current-version (pkg-tag-semver current))
+          (other-version (pkg-tag-semver other)))
+      (let lp ((current-rest current-version)
+               (other-rest other-version))
+        (match current-rest
+          ([current-hd . current-rest]
+           (match other-rest
+             ([other-hd . other-rest]
+              (cond
+               ((= current-hd other-hd)
+                (lp current-rest other-rest))
+               ((> current-hd other-hd)
+                current)
+               (else other)))
+             (else current)))
+          (else other)))))))
+
+(def +rx-semver+
+  (pregexp "v(\\d+\\.)*\\d+"))
```

Comment:
do we need the v?

Reply:
yes, semver tags have a v before.

Reply:
let's follow go's convention here, it is good.